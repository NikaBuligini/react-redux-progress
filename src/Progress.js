// @flow

import React from 'react';
import Ticker from './Ticker';

type Props = {
  isActive: boolean,
  children: (percent: number) => React$Element<*>,
};

type State = {
  progress: number,
};

class Progress extends React.PureComponent<Props, State> {
  static displayName = 'Progress';

  state = {
    progress: -1,
  };

  componentDidMount() {
    if (this.props.isActive) {
      this.updateProgress(0);
    }
  }

  componentWillUpdate(props: Props, state: State) {
    const { progress } = this.state;
    const { isActive: wasActive } = this.props;
    const { isActive } = props;

    // Start progress
    if (!wasActive && isActive) {
      this.updateProgress(0);
    }

    // Complete progress when status changes. But prevent state update while re-rendering.
    if (wasActive && !isActive && progress !== -1 && state.progress < 100) {
      this.updateProgress(100);
    }
  }

  updateProgress = (progress: number) => {
    this.setState({ progress });
  };

  render() {
    return (
      <Ticker
        {...this.props}
        percent={this.state.progress}
        updateProgress={this.updateProgress}
        renderProgress={this.props.children}
      />
    );
  }
}

export default Progress;
