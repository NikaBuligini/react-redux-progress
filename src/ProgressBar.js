// @flow

import React from 'react';

const MAX_PERCENT = 85;

type Props = {
  percent: number,
  updateProgress: (progress: number) => void,
  autoIncrement: boolean,
  intervalTime: number,
  renderProgress: (percent: number) => React$Element<*>,
};

type State = {
  percent: number,
};

class ProgressBar extends React.Component<Props, State> {
  static defaultProps = {
    percent: -1,
    autoIncrement: true,
    intervalTime: 450,
  };

  state = {
    percent: this.props.percent,
  };

  componentDidMount() {
    this.handleProps(this.props);
  }

  componentWillReceiveProps(nextProps: Props) {
    if (this.interval) {
      // stop progress when new props come in.
      clearInterval(this.interval);
      this.interval = undefined;
    }

    if (this.timeout) {
      // clear timeout when new props come in.
      clearTimeout(this.timeout);
      this.timeout = undefined;
    }

    // start progress with updated props.
    this.handleProps(nextProps);
  }

  componentWillUnmount() {
    // cleaning up interval and timeout.
    if (this.interval) {
      clearInterval(this.interval);
      this.interval = undefined;
    }

    if (this.timeout) {
      clearTimeout(this.timeout);
      this.timeout = undefined;
    }
  }

  increment = () => {
    let { percent } = this.state;

    percent += (Math.random() + 1 - Math.random()) * 3; // eslint-disable-line
    percent = percent < MAX_PERCENT ? percent : MAX_PERCENT;

    this.setState({ percent });
  };

  interval: ?NodeJS.Timer;

  handleProps = (props: Props) => {
    const { autoIncrement, percent, intervalTime } = props;

    if (autoIncrement && percent >= 0 && percent < 99) {
      this.interval = setInterval(this.increment, intervalTime);
    }

    if (percent >= 100) {
      this.setState({ percent: 99.9 }, () => {
        this.timeout = setTimeout(() => {
          this.setState({ percent: -1 }, () => this.props.updateProgress(-1));
        }, 400);
      });
    } else {
      this.setState({ percent });
    }
  };

  timeout: ?NodeJS.Timer;

  render() {
    return this.props.renderProgress(this.state.percent);
  }
}

export default ProgressBar;
