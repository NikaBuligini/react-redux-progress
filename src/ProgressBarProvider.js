// @flow

import React from 'react';

import useProgress, { usePrevious } from './useProgress';
import { nameToHex, hexToRgb } from './hexToRgb';

const DEFAULT_COLOR = '#77b6ff';

type Props = {
  isActive: boolean,
  color: string,
  className?: string,
  styles?: Object,
  absolute?: boolean,
};

const styles = {
  wrapper: {
    top: '0',
    left: '0',
    width: '100%',
    transition: 'all 500mx ease-in-out',
  },
  hiddenWrapper: {
    visibility: 'hidden',
    opacity: '0',
    zIndex: '-10',
  },
  visibleWrapper: {
    visibility: 'visible',
    opacity: '1',
    zIndex: '9999',
  },
  percent: {
    transition: 'all 400ms ease',
    height: '2px',
  },
};

function getWrapperStyles(isHidden: boolean, isAbsolute: ?boolean = false) {
  const visibilityStyles = isHidden
    ? styles.hiddenWrapper
    : styles.visibleWrapper;

  return {
    ...styles.wrapper,
    ...visibilityStyles,
    position: isAbsolute ? 'absolute' : 'fixed',
  };
}

function getPercentStyles(
  color: string,
  percent: number,
  clientStyles: Object = {},
) {
  const customStyles = {
    width: percent <= 0 ? '0%' : `${percent}%`,
    opacity: percent >= 99.9 ? '0' : '1',
  };

  return {
    ...styles.percent,
    background: color,
    boxShadow: `0 0 10px ${hexToRgb(color, 0.7)}`,
    ...clientStyles,
    ...customStyles,
  };
}

const ProgressBarProvider = ({
  isActive,
  color,
  absolute,
  className,
  styles,
  children,
}: Props) => {
  const [progressColor, setProgressColor] = React.useState(() =>
    nameToHex(color, DEFAULT_COLOR),
  );

  const [iterationKey, setIteration] = React.useState(0);

  const prevIsActive = usePrevious(isActive);

  React.useEffect(() => {
    if (isActive && !prevIsActive) {
      setIteration(prevIteration => prevIteration + 1);
    }
  }, [isActive]);

  const percent = useProgress(isActive);

  // Hide progress bar if percent is less than 0.
  const isHidden = percent < 0 || percent > 100;

  // Set `state.percent` as width.
  return (
    <div key={iterationKey} style={getWrapperStyles(isHidden, absolute)}>
      <div
        className={className}
        style={getPercentStyles(progressColor, percent, styles)}
      />
    </div>
  );
};

ProgressBarProvider.displayName = 'ProgressBarProvider';

ProgressBarProvider.defaultProps = {
  absolute: false,
  color: DEFAULT_COLOR,
  styles: {},
};

export default ProgressBarProvider;
