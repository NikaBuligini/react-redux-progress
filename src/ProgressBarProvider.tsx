import { useEffect, useState, createElement, CSSProperties } from 'react';

import { useProgress, usePrevious } from './useProgress';
import { nameToHex, hexToRgb } from './hexToRgb';

const DEFAULT_COLOR = '#77b6ff';

type Props = {
  isActive: boolean;
  color?: string;
  className?: string;
  styles?: CSSProperties;
  absolute?: boolean;
};

const styles: Record<string, CSSProperties> = {
  wrapper: {
    top: 0,
    left: 0,
    width: '100%',
    transition: 'all 500mx ease-in-out',
  },
  hiddenWrapper: {
    visibility: 'hidden',
    opacity: 0,
    zIndex: -10,
  },
  visibleWrapper: {
    visibility: 'visible',
    opacity: 1,
    zIndex: 9999,
  },
  percent: {
    transition: 'all 400ms ease',
    height: '2px',
  },
};

function getWrapperStyles(
  isHidden: boolean,
  isAbsolute: boolean = false
): CSSProperties {
  const visibilityStyles = isHidden
    ? styles.hiddenWrapper
    : styles.visibleWrapper;

  return {
    ...styles.wrapper,
    ...visibilityStyles,
    position: isAbsolute ? 'absolute' : 'fixed',
  };
}

function getPercentStyles(
  color: string,
  percent: number,
  clientStyles: Object = {}
): CSSProperties {
  const customStyles = {
    width: percent <= 0 ? '0%' : `${percent}%`,
    opacity: percent >= 99.9 ? '0' : '1',
  };

  return {
    ...styles.percent,
    background: color,
    boxShadow: `0 0 10px ${hexToRgb(color, 0.7)}`,
    ...clientStyles,
    ...customStyles,
  };
}

export const ProgressBarProvider = ({
  isActive,
  color = DEFAULT_COLOR,
  absolute = false,
  className,
  styles = {},
}: Props) => {
  const progressColor = nameToHex(color) ?? DEFAULT_COLOR;

  const [iterationKey, setIteration] = useState(0);

  const prevIsActive = usePrevious(isActive);

  useEffect(() => {
    if (isActive && !prevIsActive) {
      setIteration((prevIteration) => prevIteration + 1);
    }
  }, [isActive]); // eslint-disable-line react-hooks/exhaustive-deps

  const percent = useProgress(isActive);

  // Hide progress bar if percent is less than 0.
  const isHidden = percent < 0 || percent > 100;

  // Set `state.percent` as width.
  return createElement(
    'div',
    {
      key: iterationKey,
      style: getWrapperStyles(isHidden, absolute),
    },
    createElement('div', {
      className,
      style: getPercentStyles(progressColor, percent, styles),
    })
  );
};

ProgressBarProvider.displayName = 'ProgressBarProvider';
