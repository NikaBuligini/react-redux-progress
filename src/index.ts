export { useProgress } from './useProgress';

export type { Config } from './useProgress';
