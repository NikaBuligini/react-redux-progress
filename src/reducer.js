// @flow

type ProgressKey = string | number;
type PossibleStatuses = 'loaded' | 'loading';

type ProgressStatuses = {
  [key: ProgressKey]: PossibleStatuses,
};

type ProgressState = {
  isActive: boolean,
  fetchStatus: ProgressStatuses,
};

type ProgressAction = {
  progressId?: ProgressKey,
};

function statuses(
  state: ProgressStatuses = {},
  status: ?PossibleStatuses,
  progressId: ProgressKey,
) {
  return Object.assign({}, state, {
    [progressId]: status ? 'loaded' : 'loading',
  });
}

let activeProgresses = 0;

export function resetProgressCounter() {
  activeProgresses = 0;
}

const initialState: ProgressState = {
  isActive: false,
  fetchStatus: {},
};

export default function progress(state: ProgressState = initialState, action: ProgressAction) {
  if (action.progressId) {
    const status = state.fetchStatus[action.progressId];

    activeProgresses += status ? -1 : 1;

    return {
      isActive: activeProgresses !== 0,
      fetchStatus: statuses(state.fetchStatus, status, action.progressId),
    };
  }

  return state;
}
