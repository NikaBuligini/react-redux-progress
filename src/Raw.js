// @flow

import React from 'react';
import ProgressBar from './ProgressBar';

type Props = {
  isActive: boolean,
  renderProgress: (percent: number) => React$Element<*>,
};

type State = {
  progress: number,
};

class ProgressBarProvider extends React.PureComponent<Props, State> {
  static displayName = 'ProgressBarProvider';

  state = {
    progress: -1,
  };

  componentDidMount() {
    if (this.props.isActive) {
      this.updateProgress(0);
    }
  }

  componentWillUpdate(props: Props, state: State) {
    const { progress } = this.state;
    const { isActive: wasActive } = this.props;
    const { isActive } = props;

    // Start progress
    if (!wasActive && isActive) {
      this.updateProgress(0);
    }

    // Complete progress when status changes. But prevent state update while re-rendering.
    if (wasActive && !isActive && progress !== -1 && state.progress < 100) {
      this.updateProgress(100);
    }
  }

  updateProgress = (progress: number) => {
    this.setState({ progress });
  };

  render() {
    return (
      <ProgressBar
        {...this.props}
        percent={this.state.progress}
        updateProgress={this.updateProgress}
        renderProgress={this.props.renderProgress}
      />
    );
  }
}

export default ProgressBarProvider;
