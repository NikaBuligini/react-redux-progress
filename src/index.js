import Progress from './Progress';
import reducer from './reducer';

export { reducer };

export default Progress;
