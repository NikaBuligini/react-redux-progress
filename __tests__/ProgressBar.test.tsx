import React from 'react';
import renderer from 'react-test-renderer';

import { ProgressBarProvider } from '../src/ProgressBarProvider';

jest.useFakeTimers();

describe('<ProgressBarProvider />', () => {
  it('should render <ProgressBar />', () => {
    const component = renderer.create(<ProgressBarProvider isActive={false} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render progress bar with color name', () => {
    const component = renderer.create(
      <ProgressBarProvider isActive={false} color="red" />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render progress bar with color hex', () => {
    const component = renderer.create(
      <ProgressBarProvider isActive={false} color="#db7093" />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render progress bar with styles', () => {
    const component = renderer.create(
      <ProgressBarProvider isActive={false} styles={{ height: '5px' }} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render progress bar with className', () => {
    const component = renderer.create(
      <ProgressBarProvider isActive={false} className="my-custom-classname" />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
