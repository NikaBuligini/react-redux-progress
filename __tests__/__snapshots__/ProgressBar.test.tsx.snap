// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressBarProvider /> should render <ProgressBar /> 1`] = `
<div
  style={
    Object {
      "left": 0,
      "opacity": 0,
      "position": "fixed",
      "top": 0,
      "transition": "all 500mx ease-in-out",
      "visibility": "hidden",
      "width": "100%",
      "zIndex": -10,
    }
  }
>
  <div
    style={
      Object {
        "background": "#77b6ff",
        "boxShadow": "0 0 10px rgba(119, 182, 255, 0.7)",
        "height": "2px",
        "opacity": "1",
        "transition": "all 400ms ease",
        "width": "0%",
      }
    }
  />
</div>
`;

exports[`<ProgressBarProvider /> should render progress bar with className 1`] = `
<div
  style={
    Object {
      "left": 0,
      "opacity": 0,
      "position": "fixed",
      "top": 0,
      "transition": "all 500mx ease-in-out",
      "visibility": "hidden",
      "width": "100%",
      "zIndex": -10,
    }
  }
>
  <div
    className="my-custom-classname"
    style={
      Object {
        "background": "#77b6ff",
        "boxShadow": "0 0 10px rgba(119, 182, 255, 0.7)",
        "height": "2px",
        "opacity": "1",
        "transition": "all 400ms ease",
        "width": "0%",
      }
    }
  />
</div>
`;

exports[`<ProgressBarProvider /> should render progress bar with color hex 1`] = `
<div
  style={
    Object {
      "left": 0,
      "opacity": 0,
      "position": "fixed",
      "top": 0,
      "transition": "all 500mx ease-in-out",
      "visibility": "hidden",
      "width": "100%",
      "zIndex": -10,
    }
  }
>
  <div
    style={
      Object {
        "background": "#db7093",
        "boxShadow": "0 0 10px rgba(219, 112, 147, 0.7)",
        "height": "2px",
        "opacity": "1",
        "transition": "all 400ms ease",
        "width": "0%",
      }
    }
  />
</div>
`;

exports[`<ProgressBarProvider /> should render progress bar with color name 1`] = `
<div
  style={
    Object {
      "left": 0,
      "opacity": 0,
      "position": "fixed",
      "top": 0,
      "transition": "all 500mx ease-in-out",
      "visibility": "hidden",
      "width": "100%",
      "zIndex": -10,
    }
  }
>
  <div
    style={
      Object {
        "background": "#ff0000",
        "boxShadow": "0 0 10px rgba(255, 0, 0, 0.7)",
        "height": "2px",
        "opacity": "1",
        "transition": "all 400ms ease",
        "width": "0%",
      }
    }
  />
</div>
`;

exports[`<ProgressBarProvider /> should render progress bar with styles 1`] = `
<div
  style={
    Object {
      "left": 0,
      "opacity": 0,
      "position": "fixed",
      "top": 0,
      "transition": "all 500mx ease-in-out",
      "visibility": "hidden",
      "width": "100%",
      "zIndex": -10,
    }
  }
>
  <div
    style={
      Object {
        "background": "#77b6ff",
        "boxShadow": "0 0 10px rgba(119, 182, 255, 0.7)",
        "height": "5px",
        "opacity": "1",
        "transition": "all 400ms ease",
        "width": "0%",
      }
    }
  />
</div>
`;
